File: src/main/java/com/qtech/bubbles/util/helpers/MathHelper.java
package com.qtech.bubbles.util.helpers;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;

/**
 * Math helper, for all your math needs.
 */
public class MathHelper {
    // Clamp methods.
    @Deprecated
    public static byte clamp(byte value, byte min, byte max) {
        return (byte) Math.min(Math.max(value, min), max);
    }

    public static byte clamp(byte value, int min, int max) {
        return (byte) Math.min(Math.max(value, min), max);
    }

    @Deprecated
    public static short clamp(short value, short min, short max) {
        return (short) Math.min(Math.max(value, min), max);
    }

    public static short clamp(short value, int min, int max) {
        return (short) Math.min(Math.max(value, min), max);
    }

    public static int clamp(int value, int min, int max) {
        return Math.min(Math.max(value, min), max);
    }

    public static long clamp(long value, long min, long max) {
        return Math.min(Math.max(value, min), max);
    }

    public static float clamp(float value, float min, float max) {
        return Math.min(Math.max(value, min), max);
    }

    public static double clamp(double value, double min, double max) {
        return Math.min(Math.max(value, min), max);
    }

    public static BigInteger clamp(BigInteger value, BigInteger min, BigInteger max) {
        return value.max(min).min(max);
    }

    public static BigDecimal clamp(BigDecimal value, BigDecimal min, BigDecimal max) {
        return value.max(min).min(max);
    }

    public static double root(int value, int root) {
        return Math.pow(value, 1.0d / root);
    }

    public static double round(double value, int places) {
        if (places < 0) throw new IllegalArgumentException();

        BigDecimal bd = new BigDecimal(Double.toString(value));
        bd = bd.setScale(places, RoundingMode.HALF_UP);
        return bd.doubleValue();
    }
}

File: src/main/kotlin/qtech/bubbles/util/helpers/MathHelper.kt
package qtech.bubbles.util.helpers

import java.math.BigDecimal
import java.math.BigInteger
import java.math.RoundingMode
import kotlin.math.pow

/**
 * Math helper, for all your math needs.
 */
object MathHelper {
    // Clamp methods.
    fun clamp(value: Byte, min: Byte, max: Byte): Byte {
        return value.toInt().coerceAtLeast(min.toInt()).coerceAtMost(max.toInt()).toByte()
    }

    fun clamp(value: Byte, min: Int, max: Int): Byte {
        return value.toInt().coerceAtLeast(min).coerceAtMost(max).toByte()
    }

    fun clamp(value: Short, min: Short, max: Short): Short {
        return value.toInt().coerceAtLeast(min.toInt()).coerceAtMost(max.toInt()).toShort()
    }

    fun clamp(value: Short, min: Int, max: Int): Short {
        return value.toInt().coerceAtLeast(min).coerceAtMost(max).toShort()
    }

    fun clamp(value: Int, min: Int, max: Int): Int {
        return value.coerceAtLeast(min).coerceAtMost(max)
    }

    fun clamp(value: Long, min: Long, max: Long): Long {
        return value.coerceAtLeast(min).coerceAtMost(max)
    }

    fun clamp(value: Float, min: Float, max: Float): Float {
        return value.coerceAtLeast(min).coerceAtMost(max)
    }

    fun clamp(value: Double, min: Double, max: Double): Double {
        return value.coerceAtLeast(min).coerceAtMost(max)
    }

    fun clamp(value: BigInteger, min: BigInteger?, max: BigInteger?): BigInteger {
        return value.max(min).min(max)
    }

    fun clamp(value: BigDecimal, min: BigDecimal?, max: BigDecimal?): BigDecimal {
        return value.max(min).min(max)
    }

    @Suppress("unused")
    fun root(value: Int, root: Int): Double {
        return value.toDouble().pow(1.0 / root)
    }

    fun round(value: Double, places: Int): Double {
        require(places >= 0)
        var bd = BigDecimal(value.toString())
        bd = bd.setScale(places, RoundingMode.HALF_UP)
        return bd.toDouble()
    }
}